text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,403 @@
+- Feature Name: Arbitrary Self Types 2.0
+- Start Date: 2023-05-04
+- RFC PR: [rust-lang/rfcs#3519](https://github.com/rust-lang/rfcs/pull/3519)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Allow types that implement the new `trait Receiver<Target=Self>` to be the receiver of a method.
+
+# Motivation
+[motivation]: #motivation
+
+Today, methods can only be received by value, by reference, or by one of a few blessed smart pointer types from `core`, `alloc` and `std` (`Arc<Self>`, `Box<Self>`, `Pin<P>` and `Rc<Self>`).
+
+It's been assumed that this will eventually be generalized to support any smart pointer, such as an `CustomPtr<Self>`. Since late 2017, it has been available on nightly under the `arbitrary_self_types` feature for types that implement `Deref<Target=Self>` and for raw pointers.
+
+This RFC proposes some changes to the existing nightly feature based on the experience gained, with a view towards stabilizing the feature in the relatively near future.
+
+## Motivation for the arbitrary self types feature overall
+
+One use-case is cross-language interop (JavaScript, Python, C++), where other languages' references can’t guarantee the aliasing and exclusivity semantics required of a Rust reference. For example, the C++ `this` pointer can't be practically or safely represented as a Rust reference because C++ may retain other pointers to the data and it might mutate at any time. Yet, calling C++ methods intrinsically requires a `this` reference.
```

Comment:
This needs expanding, since it sounds like arbitrary self types are mainly useful for unsafe code interacting with FFI? That would be very niche.

I think what you are hinting at here is that such code could define a newtype around a raw pointer and then provide a completely safe API? If so, then please don't just hint at this -- actually spell out an example, showing the desired Rust API.

Reply:
Added - thanks - you're right that we're aiming to use this from safe Rust code.